---
title: "SSHAARP: Searching Shared HLA Amino Acid Residue Prevalence"
author: "Livia Tran - livtran@chori.org"
date: "December 4, 2019"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SSHAARP Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview
The Searching Shared HLA Amino Acid Residue Prevalence (SSHAARP) Package processes amino acid alignments produced by the IPD-IMGT/HLA Database to identify user-defined amino acid residue motifs shared across HLA alleles, and calculates the frequencies of those motifs based on HLA allele frequency data. SSHAARP uses Generic Mapping Tools (GMT) software and the GMT R Package to generate global frequency heat maps that illustrate the distribution of each user-defined map around the globe. SSHAARP analyzes the allele frequency data described in Solberg et al. 2008, a global set of 497 population samples from 185 published datasets, representing 66,800 individuals.

Future plans for SSHAARP include producing heat maps based on allelic frequencies obtained from the Allele Frequency Net Database (AFND). 

SSHAARP obtains HLA amino acid alignments from the IPD-IMGT/HLA GitHub repository at: https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments.

The required Solberg dataset (See: doi: 10.1016/j.humimm.2008.05.001) file is the 1-locus-alleles.dat file in the results.zip archive at http://pypop.org/popdata/.

SSHAARP's functions interact with two external datasets and one R package:

1. IPD-IMGT/HLA:

    **BLAASD()** extracts alignment sequence information for a given locus from the ANHIG/IMGTHLA Github repository to produce a dataframe of amino acid sequence data for each individual peptide position for all alleles.
    
      **findMotif** isolates the individual amino acid position dataframe produced by BLAASD() to only alleles with the user-defined motif. 
      
2. Solberg Dataset:

      **convertCoordinates()** converts coordinates in the Solberg dataset with cardinal directions to their appropriate enumerations (i.e. 50S is converted to -50)
  
      **dataSubset()()** manipulates the Solberg dataset by adding in a new column containing locus*allele information, reordering the dataset based on population name, and subsetting the data to only the locus of interest.
  
3. Generic Mapping Tools (GMT)

      **PALM()()** applies the Generic Mapping Tools (gmt) R package to the modified Solberg data to produce heatmaps. The gmt R package is an interface between R and the GMT Map-Making software, and is required for the PALM()() function. GMT commands are called via a Bash script through using the gmt.system() function in the GMT R package. Both the gmt R package **AND** GMT software must be installed to run SSHAARP, which is available here: https://www.soest.hawaii.edu/gmt/.
  
---

## Functions

### BLAASD()
BLAASD() extracts alignment sequence information for a given locus from the ANHIG/IMGTHLA Github Repository to produce a dataframe, which contains individual amino acid data for each amino acid position for all alleles, for a user-defined HLA locus or loci. The first 4 columns are locus, allele, trimmed allele, and allele_name.

Most DQ-Beta proteins (e.g., `DQB1*05:01`) lack the eight amino-acids encoded by exon 5 due to an intron 4 splice-acceptor mutation. Some DQB1 alleles (e.g., `DQB1*05:03` and `*06:01`) do not have this intron mutation, and include these eight amino acids. However, because `DQB1*05:01` is the DQB1 reference allele, these amino acid positions are represented as insertion positions in the protein alignment generated by the IPD-IMGT/HLA Database. It is not clear how these exon 5-encoded amino acid positions are numbered in the IPD-IMGT/HLA Database; BLAASD() assigns amino-acid position numbers 227-234 to these eight amino-acid indel positions, and the C-terminal DQB1 amino-acid position in the DQB1 protein alignment generated by BLAASD() is position 237.

Example output of BLAASD() for HLA-A, where the selected row outputs are 2, 549, and 800 to illustrate different HLA-A alleles, and selected columns are the first 20 columns of the dataframe, where the first 4 columns consist of the locus, allele, trimmed_allele of 2 fields, and the full allele name. The rest of the columns consist of corresponding amino acids for each amino acid position for a given allele. 

```
>BLAASD()("A")[[1]][c(2,8),1:20]
    locus      allele trimmed_allele   allele_name 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16
2       A 01:01:01:01        A*01:01 A*01:01:01:01 M A V M A P R T L  L  L  L  L  S  G  A
549     A    02:01:30        A*02:01    A*02:01:30 M A V M A P R T L  V  L  L  L  S  G  A
800     A    02:30:01        A*02:30    A*02:30:01 M A V M A P R T L  V  L  L  L  S  G  A
```


The first row of BLAASD() is a correspondence table of amino acid position enumerations. Alignments begin with a negative enumeration for the the amino acid positions within the first exon, which encodes the leader peptide, since the leader peptide is eventually cleaved off. The rest of the amino acid positions associated with exons other than exon 1 have positive enumerations. In addition, the alignment counts do not give amino acid positions with inDels a count. The correspondence table was created to begin enumeration at position 1, since it is more standard, and to account for the skipped over positions with inDels. Below is an example of what the correspondence table looks like -- the alignment sequence for the first amino acid postion, -24, corresponds to a more standard beginning enumeration of 1.

```
>BLAASD()("A")[[1]][1,1:20]
  locus allele trimmed_allele allele_name   1   2   3   4   5   6   7   8   9  10  11  12
1  <NA>   <NA>           <NA>        <NA> -24 -23 -22 -21 -20 -19 -18 -17 -16 -15 -14 -13
   13  14  15 16
1 -12 -11 -10 -9
###
```

### findMotif()
findMotif() splits up the input motif (should be in format Locus*AA_1*AA_2*AA_3, where the amino acid positions are in numerical order). If the motif is not in numerical order, such as DRB1*30Y~28E~26F, the function will rearrange the motif such that the 26F is searched for first. The following output is an example of the 1st, 5th, and 9th DRB1 alleles that have the DRB1*26F~28E~30Y motif. Recall the dataframe is formatted with actual sequence enumeration (starting with 1), but the search is done by matching the input motif, which is based off of alignment enumeration, with the correspondence table. Thus, in the output dataframe, alignment position 26 is equivalent to actual position 55, 28 to 58, and 30 to 60.

```
>findMotif()("DRB1*26F~28E~30Y")[c(1,5,9,10), 1:65]

    locus   allele trimmed_allele   allele_name   1   2   3   4   5   6   7   8   9  10  11
1    <NA>     <NA>           <NA>          <NA> -29 -28 -27 -26 -25 -24 -23 -22 -21 -20 -19
197  DRB1 03:02:04     DRB1*03:02 DRB1*03:02:04   *   *   *   *   *   *   *   *   *   *   *
240  DRB1    03:38     DRB1*03:38    DRB1*03:38   *   *   *   *   *   *   *   *   *   *   *
256  DRB1    03:53     DRB1*03:53    DRB1*03:53   *   *   *   *   *   *   *   *   *   *   *
     12  13  14  15  16  17  18  19  20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37
1   -18 -17 -16 -15 -14 -13 -12 -11 -10 -9 -8 -7 -6 -5 -4 -3 -2 -1  1  2  3  4  5  6  7  8
197   *   *   *   *   *   *   *   *   *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  R  F  L
240   *   *   *   *   *   *   *   *   *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  R  F  L
256   *   *   *   *   *   *   *   *   *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  R  F  L
    38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55      56 57 58 59 60 61
1    9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 InDel_1 27 28 29 30 31
197  E  Y  S  T  S  E  C  H  F  F  N  G  T  E  R  V  R  F       .  L  E  R  Y  F
240  E  Y  S  T  S  E  C  H  F  F  N  G  T  E  R  V  R  F       .  L  E  R  Y  F
256  E  Y  S  T  S  E  C  H  F  F  N  G  T  E  R  V  R  F       .  L  E  R  Y  F
```

If a motif is not found, an error message is output:

```
>findMotif()("DRB1*26F~28E~30Z")

Warning message:
In findMotif()("DRB1*26F~28E~30Z") :
  Error - zero alleles match this motif. Please try again.
```

### dataSubset()
dataSubset() manipulates the Solberg dataset to prepare for heatmap generation by adding in a new column containing locus*allele information, reordering the dataset based on population name, and subsetting the dataset to only the locus of interest.

```
Partial depiction of Solberg Dataset before dataSubset() - note population names are out of order, and the data consists of data from multiple HLA loci

>dataset[c(1,2,300,600, 1000),]

     dataset     popname contin complex   latit  longit           coord locus allele_v2
1        lit Korean_1999    NEA       3  37.00N 127.30E  37.00N 127.30E  DRB1      1302
2        lit  Gabes_2006    NAF       2     34N     10E         34N 10E  DRB1      1311
300     ihwg      Ami_97    SEA       1 25.067N 121.65E 25.067N 121.65E     B      5601
600     ihwg   Brazilian    OTH    3mig     10S     55W         10S 55W     B      5001
1000    ihwg     Chaouya    NAF       2   35.5N      8W        35.5N 8W     A      2902
     allele_v3 allele.freq allele.count gametes
1        13:02     0.11001          111    1009
2        13:11     0.00526            1     190
300      56:01     0.20408           40     196
600      50:01     0.03371            6     178
1000     29:02     0.06716            9     134
```

```
Partial depiction of Solberg Dataset after dataSubset() - note population names are now in lexicographic order, the only locus of note in the dataset is the locus provided in the motif (i.e. DRB1), and there is a new column, locus_allele, for comparison to the dataframe produced by BLAASD().

>dataSubset()(dataset, "DRB1*26F~28E~30Y")[c(1,2,300,600, 1000),]

      dataset        popname contin complex  latit longit         coord locus allele_v2
8615      lit      Ache_2003    SAM       1    24S    56W       24S 56W  DRB1      0403
8616      lit      Ache_2003    SAM       1    24S    56W       24S 56W  DRB1      0411
9365      lit      Bari_1994    SAM       2    10N    70W       10N 70W  DRB1      0403
10068     lit Cameroon_2001b    SSA       3  6.00N 12.00E  6.00N 12.00E  DRB1      1102
10679     lit     Czech_1992    EUR       3 49.45N 15.30E 49.45N 15.30E  DRB1      0301
      allele_v3 allele.freq allele.count gametes locus_allele
8615      04:03     0.00578            1     173   DRB1*04:03
8616      04:11     0.74566          129     173   DRB1*04:11
9365      04:03     0.03636            4     110   DRB1*04:03
10068     11:02     0.03175            8     252   DRB1*11:02
10679     03:01     0.11111           21     189   DRB1*03:01
```

### convertCoordinates()
convertCoordinates() converts coordinates in the Solberg dataset with a cardinal direction to its appropriate numerical enumeration for plotting on a map. North (N), which is above the equator, and East (E), which is east of the prime meridian are positive, while South (S), below the equator, and West (W), west of the prime meridian, are negative. 

```
Partial depiction of Solberg Dataset before convertCoordinates

                   popname contin latit longit
8615             Ache_2003    SAM   24S    56W
101            Algerian_99    NAF   35N   0.5W
17938 BrazilGuaraniNandeva    SAM   23S    54W
```

```
Partial depiction of Solberg Dataset after convertCoordinates

>convertCoordinates(heatmapdata)

                   popname contin latit longit
8615             Ache_2003    SAM   -24    -56
101            Algerian_99    NAF   35     -0.5
17938 BrazilGuaraniNandeva    SAM   -23    -54
```

### PALM() 

PALM() produces a heatmap from the manipulated Solberg dataset by using the Generic Mapping Tools (GMT) R Package, which is an interface between R and the GMT Map-Making software. GMT software is required for this function and can be downloaded at https://www.soest.hawaii.edu/gmt/. GMT commands are called via a Bash script through using the gmt.system() function in the GMT R package. The final output of PALM() is a heatmap saved as a JPEG in the user's working directory. 

Two default parameters are set for PALM():
  1) color = TRUE
  Should the heatmap produced be in color? The default for color is set to true for a color heatmap – for a map in grayscale, set color==FALSE.
  2) filter_migrant = TRUE
  Should OTH and migrant populations be filtered out of the dataset? The default for filter_migrant is set to true – set this parameter equal to FALSE to keep migrant and OTH populations in the dataset.

```
#JPEG with color
>PALM(Solberg_datset, "DRB1*26F~28E~30Y" )
```

```{r, out.width='100%', fig.align='center', fig.cap='...'}
knitr::include_graphics("basemap_color.jpg")
```
  
```
#JPEG without color
>PALM(Solberg_datset, "DRB1*26F~28E~30Y", color=FALSE)
```

```{r, out.width='100%', fig.align='center', fig.cap='...'}
knitr::include_graphics("basemap_bw.jpg")
```
  



